File: problem_5.py
from math import ceil
from functools import reduce
from collections import Counter

def isprime(n):
    """Is n prime?"""
    if n == 2:
        return True

    for i in range(2, ceil(n**0.5)+1):
        if n % i == 0:
            return False
    return True


def calc_arity(n, factor):
    """How many times is n divisible by factor?"""
    exp = 0
    while int(n % factor**(exp+1)) == 0:
        exp += 1
    return exp


def factorize(n):
    """Factorize n into its prime factors"""
    if isprime(n):
        return [(n,1)]

    prime_factors = []
    for factor in range(2, ceil(n/2)+1):
        if n % factor == 0 and isprime(factor):
            prime_factors.append((factor, calc_arity(n, factor)))
    return prime_factors


def add_decompositions(decomp1, decomp2):
    """Add two prime-factor decompositions by max(arity1, arity2) for each factor."""
    result = []

    d1 = Counter(dict(decomp1))
    d2 = Counter(dict(decomp2))

    for factor in set(d1).union(set(d2)):
        result.append( (factor, max(d1[factor], d2[factor])) )

    return result


if __name__ == "__main__":
    N = 20
    decompositions = [factorize(i) for i in range(2, N + 1)]

    added = reduce(lambda d1, d2: add_decompositions(d1, d2), decompositions)

    product = 1
    for factor, power in added:
        product *= factor ** power

    print(f"The smallest number evenly divisible by 1 to {N} is {product}")

File: problem_5.cpp
#include <stdio.h>
#include <math.h>
#include <vector>
#include <algorithm>

/**
 * Euler project
 *   problem 5 -- Smallest multiple
 *
 * 2520 is the smallest number that can be divided by each of the numbers from
 * 1 to 10 without any remainder. What is the smallest positive number that is
 * evenly divisible by all of the numbers from 1 to 20?
 */

typedef std::vector< std::pair<long, long>> PrimeDecomposition;

int isprime(long n) {
    for(int i=2; i <= sqrt(n); i++) {
        if (n % i == 0) {
            return 0;
        }
    }
    return 1;
}

int calc_arity_of_divisor(long n, long divisor) {
    int arity = 0;
    long div = divisor;
    while (n % div == 0) {
        arity++;
        div *= divisor;
    }
    return arity;
}

auto decompose_into_prime_factors(long n) {
    std::vector< std::pair<long,long> > factors;

    if (isprime(n)) {
        factors.push_back( std::make_pair( n, 1 ));
        return factors;
    }

    for(int factor=2; factor <= ceil(n/2); factor++) {
        if ((n % factor == 0) && isprime(factor)) {
            factors.push_back( std::make_pair( factor, calc_arity_of_divisor(n, factor)));
        }
    }

    return factors;
}

/**
 * This, in fact, finds the smallest common multiple of both d1 and d2.
 */
auto add_prime_decompositions(PrimeDecomposition d1, PrimeDecomposition d2) {
    std::vector< std::pair<long,long> > result;

    for(auto &p1 : d1) {
        long factor = p1.first;
        long power = p1.second;

        // update existing ones (those in both d1 and d2)
        auto found = std::find_if(d2.begin(), d2.end(),
                        [factor] (std::pair<long,long> p) {return p.first==factor;}
                     );
        if (found != d2.end() && found->second > power) {
            power = found->second;
        }

        result.push_back(std::make_pair(factor, power));
    }

    // add in the rest (those in d2 but not in d1)
    for(auto &p2 : d2) {
        if (d1.end() == std::find_if(d1.begin(), d1.end(),
                                     [p2] (std::pair<long,long> p1)
                                          {return p1.first == p2.first;}
                                    ))
        {
            result.push_back(p2);
        }
    }

    // sort and return
    std::sort(result.begin(), result.end());

    return result;
}

int main() {

    // sum up the prime factors with their multiplicities
    PrimeDecomposition added;
    for(int i=1; i<=20; i++) {
        auto d = decompose_into_prime_factors(i);
        added = add_prime_decompositions(added, d);
    }

    // get the result
    long product = 1;
    for(auto &p : added) {
        product *= (long)pow(p.first, p.second);
    }

    printf("The smallest number evenly divisible by 1 to 20 is %ld\n", product);
    return 0;
}
